File: app/src/main/java/arttttt/rotationcontrol/RotationService.java
package arttttt.rotationcontrol;

import android.app.Notification;
import android.app.Notification.Builder;
import android.app.PendingIntent;
import android.app.Service;
import android.content.ContentResolver;
import android.content.Intent;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.provider.Settings;
import android.view.WindowManager;
import android.widget.RemoteViews;
import android.provider.Settings.System;

import com.miui.rotationcontrol.R;

public class RotationService extends Service {

    public static final int ORIENTATION_AUTO = -1;
    private static final int ORIENTATION_PORTRAIT = 0;
    private static final int ORIENTATION_LANDSCAPE = 1;
    private static final int ORIENTATION_PORTRAIT_REVERSE = 2;
    private static final int ORIENTATION_LANDSCAPE_REVERSE = 3;

    private static final int ACTION_CLICK = 2;

    private static final String INTENT_CLICKED_BUTTON_ID = "BUTTON_ID";
    private static final String INTENT_ACTION = "ACTION";

    private String pref_orientation;

    private Notification mNotification;
    private RemoteViews mNotificationContent;
    private int mScreenOrientation;
    private RotationSettings mRotationSettings;
    private AccelerometerRotationObserver mAccelerometerRotationObserver;

    private static final int NOTIFICATION_ID = 999;
    private final int[] IDS_BUTTON = new int[]{R.id.btn_auto, R.id.btn_portrait, R.id.btn_portrait_reverse, R.id.btn_landscape, R.id.btn_landscape_reverse};

    @Override
    public void onCreate() {
        super.onCreate();

        pref_orientation = getResources().getString(R.string.prefkey_orientation);

        mRotationSettings = new RotationSettings(this);

        mScreenOrientation = Integer.valueOf(mRotationSettings.getString(pref_orientation, String.valueOf(((WindowManager)getSystemService(WINDOW_SERVICE)).getDefaultDisplay().getOrientation())));

        mAccelerometerRotationObserver = new AccelerometerRotationObserver(this, new Handler());

        createNotification();
        updateNotification();
        setOrientation();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        getContentResolver().unregisterContentObserver(mAccelerometerRotationObserver);
    }

    private void onClick(int id){
        switch (id){
            case R.id.btn_auto:{
                mScreenOrientation = ORIENTATION_AUTO;
                break;
            }
            case R.id.btn_portrait:{
                mScreenOrientation = ORIENTATION_PORTRAIT;
                break;
            }
            case R.id.btn_portrait_reverse:{
                mScreenOrientation = ORIENTATION_PORTRAIT_REVERSE;
                break;
            }
            case R.id.btn_landscape: {
                mScreenOrientation = ORIENTATION_LANDSCAPE;
                break;
            }
            case R.id.btn_landscape_reverse:{
                mScreenOrientation = ORIENTATION_LANDSCAPE_REVERSE;
                break;
            }
        }
        setOrientation();
        updateNotification();
        saveOrientation();
    }

    private void setOrientation(){
        ContentResolver contentResolver = getContentResolver();
        mAccelerometerRotationObserver.onOrientationChanged(mScreenOrientation);
        System.putInt(contentResolver, "user_rotation", mScreenOrientation);
    }

    private void createNotification(){
        if (this.mNotification == null) {
            mNotificationContent = new RemoteViews(getPackageName(), R.layout.notification);
            PendingIntent activity = PendingIntent.getActivity(this, 0, new Intent(this, PreferenceActivity.class), Intent.FILL_IN_COMPONENT);

            Intent intent = new Intent(this, RotationService.class);
            intent.putExtra(INTENT_ACTION, ACTION_CLICK);
            for (int length = IDS_BUTTON.length, i = 0; i < length; ++i) {
                int id = IDS_BUTTON[i];
                intent.putExtra(INTENT_CLICKED_BUTTON_ID, id);
                mNotificationContent.setOnClickPendingIntent(id, PendingIntent.getService(this, id, intent, PendingIntent.FLAG_UPDATE_CURRENT));
            }

            int smallIconId = -1;
            switch (mScreenOrientation){
                case ORIENTATION_AUTO:{
                    smallIconId = R.drawable.ic_stat_notify_auto;
                    break;
                }
                case ORIENTATION_PORTRAIT: {
                    smallIconId = R.drawable.ic_stat_notify_portrait;
                    break;
                }
                case ORIENTATION_PORTRAIT_REVERSE:{
                    smallIconId = R.drawable.ic_stat_notify_portrait_reverse;
                    break;
                }
                case ORIENTATION_LANDSCAPE: {
                    smallIconId = R.drawable.btn_landscape;
                    break;
                }
                case ORIENTATION_LANDSCAPE_REVERSE: {
                    smallIconId = R.drawable.btn_landscape_reverse;
                    break;
                }
            }

            this.mNotification = new Builder(this).
                    setContent(mNotificationContent).
                    setContentIntent(activity).
                    setSmallIcon(smallIconId).
                    setContentText("TEST TEXT").
                    setPriority(Notification.PRIORITY_MAX).
                    build();

            showNotification();
        }
    }

    private void saveOrientation(){
        mRotationSettings.changePreference(pref_orientation, mScreenOrientation);
    }

    private void updateNotification(){
        for (int length = IDS_BUTTON.length, i = 0; i < length; ++i) {
            int id = IDS_BUTTON[i];
            mNotificationContent.setInt(id, "setBackgroundResource", android.R.color.transparent);
            mNotificationContent.setInt(id, "setColorFilter", getResources().getInteger(R.integer.color_background_inactive_filter));
        }
        int buttonId = -1;
        int modeDescriptionId = -1;
        int smallIconId = -1;
        switch (mScreenOrientation){
            case ORIENTATION_AUTO:{
                buttonId = R.id.btn_auto;
                smallIconId = R.drawable.ic_stat_notify_auto;
                modeDescriptionId = R.string.auto_mode;
                break;
            }
            case ORIENTATION_PORTRAIT: {
                buttonId = R.id.btn_portrait;
                modeDescriptionId = R.string.portrait_mode;
                smallIconId = R.drawable.ic_stat_notify_portrait;
                break;
            }
            case ORIENTATION_PORTRAIT_REVERSE:{
                buttonId = R.id.btn_portrait_reverse;
                modeDescriptionId = R.string.portrait_reverse_mode;
                smallIconId = R.drawable.ic_stat_notify_portrait_reverse;
                break;
            }
            case ORIENTATION_LANDSCAPE: {
                buttonId = R.id.btn_landscape;
                modeDescriptionId = R.string.landscape_mode;
                smallIconId = R.drawable.btn_landscape;
                break;
            }
            case ORIENTATION_LANDSCAPE_REVERSE: {
                buttonId = R.id.btn_landscape_reverse;
                modeDescriptionId = R.string.landscape_reverse_mode;
                smallIconId = R.drawable.btn_landscape_reverse;
                break;
            }
        }
        mNotificationContent.setInt(buttonId, "setBackgroundResource", R.drawable.btnback_active);
        mNotificationContent.setInt(buttonId, "setColorFilter", android.R.color.transparent);
        mNotificationContent.setInt(R.id.rotationMode, "setText", modeDescriptionId);
        mNotification.icon = smallIconId;

        showNotification();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            int action = intent.getIntExtra(INTENT_ACTION, 0);
            switch (action){
                case ACTION_CLICK:
                    onClick(intent.getIntExtra(INTENT_CLICKED_BUTTON_ID, 0));
            }
        }
        else{
            updateNotification();
        }
        return START_STICKY;
    }

    private void showNotification(){
        startForeground(NOTIFICATION_ID, mNotification);
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/arttttt/rotationcontrolv3/device/services/RotationService.kt
@file:Suppress("RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")

package com.arttttt.rotationcontrolv3.device.services

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import android.view.Surface
import android.view.WindowManager
import android.widget.RemoteViews
import androidx.core.app.NotificationCompat
import com.arttttt.rotationcontrolv3.R
import com.arttttt.rotationcontrolv3.device.services.base.BaseService
import com.arttttt.rotationcontrolv3.device.services.di.rotationServiceModule
import com.arttttt.rotationcontrolv3.device.services.helper.IRotationServiceHelper
import com.arttttt.rotationcontrolv3.utils.OsUtils
import com.arttttt.rotationcontrolv3.utils.SAVED_ORIENTATION
import com.arttttt.rotationcontrolv3.utils.delegates.permissions.ICanWriteSettingsChecker
import com.arttttt.rotationcontrolv3.utils.delegates.preferences.IPreferencesDelegate
import com.arttttt.rotationcontrolv3.utils.extensions.android.*
import com.arttttt.rotationcontrolv3.utils.extensions.koilin.unsafeCastTo
import com.arttttt.rotationcontrolv3.utils.rxjava.ISchedulersProvider
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import org.koin.core.KoinComponent
import org.koin.core.context.loadKoinModules
import org.koin.core.context.unloadKoinModules
import org.koin.core.inject

class RotationService: BaseService() {
    companion object ServiceHelper: IRotationServiceHelper, KoinComponent {
        private val context: Context by inject()
        private val permissionsChecker: ICanWriteSettingsChecker by inject()
        private val schedulers: ISchedulersProvider by inject()
        private val serviceStatus = BehaviorSubject.createDefault(IRotationServiceHelper.Status.STOPPED)

        override fun getStatusObservable(): Observable<IRotationServiceHelper.Status> = serviceStatus.serialize().share()

        override fun startRotationService() {
            permissionsChecker
                .canWriteSettings()
                .subscribeOn(schedulers.computation)
                .observeOn(schedulers.ui)
                .subscribeUntilDestroy { canWriteSettings -> dispatchCanWriteSettings(canWriteSettings) }
        }

        override fun stopRotationService() {
            stopService<RotationService>(context)
        }

        private fun dispatchCanWriteSettings(canWriteSettings: Boolean) {
            if (canWriteSettings) {
                context.startForegroundServiceCompat(intentOf<RotationService>(context))
            } else {
                toastOf(context, R.string.can_not_write_settings_toast)
            }
        }

        private const val INTENT_ACTION = "ACTION"
        private const val INTENT_ACTION_CLICK = 2
        private const val INTENT_CLICKED_BUTTON_ID = "BUTTON_ID"

        private const val NOTIFICATION_CHANNEL = "Rotation Control"
        private const val NOTIFICATION_CHANNEL_ID = "rotation_service_notification_channel_id"
        private const val NOTIFICATION_ID = 999

        private const val ORIENTATION_PORTRAIT = Surface.ROTATION_0
        private const val ORIENTATION_LANDSCAPE = Surface.ROTATION_90
        private const val ORIENTATION_PORTRAIT_REVERSE = Surface.ROTATION_180
        private const val ORIENTATION_LANDSCAPE_REVERSE = Surface.ROTATION_270
        private const val ORIENTATION_AUTO = 4
    }

    private val notificationService: NotificationManager by lazy { getSystemService(Context.NOTIFICATION_SERVICE).unsafeCastTo<NotificationManager>() }
    private val windowService: WindowManager by lazy { getSystemService(Context.WINDOW_SERVICE).unsafeCastTo<WindowManager>() }

    private var currentOrientation = ORIENTATION_AUTO

    private val buttonsIds = arrayOf(
        R.id.btn_auto,
        R.id.btn_portrait,
        R.id.btn_portrait_reverse,
        R.id.btn_landscape,
        R.id.btn_landscape_reverse
    )

    private val accelerometerObserver: AccelerometerObserver by inject()
    private val preferencesDelegate: IPreferencesDelegate by inject()

    override fun onCreate() {
        loadKoinModules(rotationServiceModule)
        super.onCreate()
        initForegroundService()
        serviceStatus.onNext(IRotationServiceHelper.Status.STARTED)
    }

    override fun onDestroy() {
        super.onDestroy()
        serviceStatus.onNext(IRotationServiceHelper.Status.STOPPED)
        unloadKoinModules(rotationServiceModule)
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent == null) return START_STICKY

        val action by intent.extra(INTENT_ACTION, 0)

        when(action) {
            INTENT_ACTION_CLICK -> {
                val buttonId by intent.extra(INTENT_CLICKED_BUTTON_ID, 0)
                handleNotificationClick(buttonId)
            }
        }

        return START_STICKY
    }

    private fun initForegroundService() {
        accelerometerObserver
            .getAccelerometerObservable()
            .subscribeOn(schedulers.computation)
            .observeOn(schedulers.computation)
            .flatMapSingle { permissionsChecker.canWriteSettings() }
            .filter { canWriteSettings -> canWriteSettings }
            .filter { currentOrientation != ORIENTATION_AUTO }
            .subscribeUntilDestroy { dispatchAccelerometerChanges() }

        contentResolver.registerContentObserver(
            Settings.System.ACCELEROMETER_ROTATION,
            false,
            accelerometerObserver
        )

        setOrientation(getInitialRotation())
            .subscribeOn(schedulers.computation)
            .observeOn(schedulers.ui)
            .subscribeUntilDestroy {
                createNotificationChannelIfNeeded()

                showNotification(
                    createNotification(
                        orientationToButtonId(currentOrientation)
                    )
                )
            }
    }

    private fun dispatchAccelerometerChanges() {
        contentResolver.putInt(Settings.System.ACCELEROMETER_ROTATION, 0)
    }

    private fun createNotification(activeButtonId: Int): Notification {
        val notificationContent = RemoteViews(packageName, R.layout.notification)

        buttonsIds.forEach { buttonId ->
            notificationContent.setOnClickPendingIntent(buttonId, PendingIntent.getService(
                context,
                buttonId,
                intentOf<RotationService>(context) {
                    putExtra(INTENT_ACTION, INTENT_ACTION_CLICK)
                    putExtra(INTENT_CLICKED_BUTTON_ID, buttonId)
                },
                PendingIntent.FLAG_UPDATE_CURRENT)
            )

            configureNotificationButton(notificationContent, buttonId, activeButtonId)
        }

        return notificationOf(this, NOTIFICATION_CHANNEL_ID) {
            setCustomContentView(notificationContent)
            setSmallIcon(R.drawable.ic_rotate)
            priority = NotificationCompat.PRIORITY_MAX
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                setStyle(NotificationCompat.DecoratedCustomViewStyle())
            }
        }.apply { flags = NotificationCompat.FLAG_ONLY_ALERT_ONCE }
    }

    private fun configureNotificationButton(notificationContent: RemoteViews, buttonId: Int, activeButtonId: Int) {
        notificationContent.setBackgroundResource(
            buttonId,
            if (buttonId == activeButtonId) {
                R.drawable.active_button_background
            } else {
                android.R.color.transparent
            }
        )

        notificationContent.setColorFilter(
            buttonId,
            context.getColorCompat(
                if (buttonId == activeButtonId) {
                    if (OsUtils.isMiui())
                        R.color.colorActiveButtonTintMiui
                    else
                        R.color.colorActiveButtonTint
                } else {
                    if (OsUtils.isMiui())
                        R.color.colorInactiveButtonTintMiui
                    else
                        R.color.colorInactiveButtonTint
                }
            )
        )
    }

    private fun handleNotificationClick(buttonId: Int) {
        permissionsChecker
            .canWriteSettings()
            .subscribeOn(schedulers.computation)
            .subscribeOn(schedulers.computation)
            .filter { canWriteSettings -> canWriteSettings }
            .flatMapSingleElement { setOrientation(buttonIdToOrientation(buttonId)).toSingleDefault(Unit) }
            .subscribeUntilDestroy(
                onComplete = {
                    toastOf(context, R.string.can_not_write_settings_toast)

                    stopSelf()
                },
                onNext =  { showNotification(createNotification(orientationToButtonId(currentOrientation))) }
            )
    }

    private fun setOrientation(newOrientation: Int): Completable {
        return permissionsChecker
            .canWriteSettings()
            .subscribeOn(schedulers.computation)
            .observeOn(schedulers.computation)
            .filter { canWriteSettings -> canWriteSettings }
            .doOnSuccess {
                if (newOrientation == ORIENTATION_AUTO) {
                    contentResolver.putInt(Settings.System.ACCELEROMETER_ROTATION, 1)
                } else {
                    dispatchAccelerometerChanges()
                    contentResolver.putInt(Settings.System.USER_ROTATION, newOrientation)
                }
                currentOrientation = newOrientation
                preferencesDelegate.putInt(SAVED_ORIENTATION, currentOrientation)
            }
            .ignoreElement()
    }

    private fun showNotification(notification: Notification) {
        startForeground(NOTIFICATION_ID, notification)
    }

    private fun createNotificationChannelIfNeeded() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            notificationService.createNotificationChannel(
                NotificationChannel(
                    NOTIFICATION_CHANNEL_ID,
                    NOTIFICATION_CHANNEL,
                    NotificationManager.IMPORTANCE_HIGH
                )
            )
        }
    }

    private fun getInitialRotation(): Int {
        return preferencesDelegate.getInt(SAVED_ORIENTATION) ?: windowService.defaultDisplay.rotation
    }

    private fun buttonIdToOrientation(buttonId: Int): Int {
        return when (buttonId) {
            R.id.btn_portrait -> ORIENTATION_PORTRAIT
            R.id.btn_portrait_reverse -> ORIENTATION_PORTRAIT_REVERSE
            R.id.btn_landscape -> ORIENTATION_LANDSCAPE
            R.id.btn_landscape_reverse -> ORIENTATION_LANDSCAPE_REVERSE
            else -> ORIENTATION_AUTO
        }
    }

    private fun orientationToButtonId(orientation: Int): Int {
        return when (orientation) {
            ORIENTATION_PORTRAIT -> R.id.btn_portrait
            ORIENTATION_PORTRAIT_REVERSE -> R.id.btn_portrait_reverse
            ORIENTATION_LANDSCAPE -> R.id.btn_landscape
            ORIENTATION_LANDSCAPE_REVERSE -> R.id.btn_landscape_reverse
            else ->  R.id.btn_auto
        }
    }
}